JAVA-2684: Add documentation and examples of Quarkus Cassandra extension (#9)
datastax/cassandra-quarkus: documentation/src/main/asciidoc/cassandraclient.adoc

= Quarkus - Using the Cassandra Client

> Cassandra is a well known NoSQL Database that is widely used.

Reviewer: This is a bit vague. Even if this is just a brief introductory description, we could be more specific. The first sentence of the Wikipedia article sounds like a good starting point:

> Apache Cassandra is a free and open-source, distributed, wide column store, NoSQL database management system designed to handle large amounts of data across many commodity servers, providing high availability with no single point of failure.

In this guide, we will see how you can get your REST services to use a Cassandra database.

== Prerequisites

To complete this guide, you need:

* an IDE
* JDK 1.8+ installed with `JAVA_HOME` configured appropriately
* GraalVM installed with `GRAALVM_HOME` configured appropriately if you want to use the native mode.
* Apache Maven 3.5.3+
* Cassandra installed or Docker installed

== Architecture

The application built in this guide is quite simple: the user can add elements in a list using a form and the list is updated.

All the information between the browser and the server is formatted as JSON.

The elements are stored in the Cassandra Database.

== Solution

We recommend that you follow the instructions in the next sections and create the application step by step.
However, you can go right to the completed example.

The solution is located in the `quickstart` {quickstart-clone-url}[directory].

== Creating the Maven project

First, we need a new project. Create a new maven project and copy the `pom.xml` file that is present in the `quickstart` directory.

The `pom.xml` is importing the RESTEasy/JAX-RS, JSON-B, Context Propagation and Cassandra Client extensions.

> We will be building a REST application using the DataStax Object Mapper {mapper-docs-url}[DataStax Object Mapper].

Reviewer: ```suggestion
We will be building a REST application using the {mapper-docs-url}[DataStax Object Mapper].
```

It will simplify the Data Access Layer code.

The most important part of the `pom.xml` is adding the `cassandra-quarkus` extension:
[source,xml]
----
<dependency>
  <groupId>com.datastax.oss.quarkus</groupId>
  <artifactId>cassandra-quarkus-client</artifactId>
  <version>${quarkus.version}</version>
</dependency>

> ----

Reviewer: Reminder to maybe update this if we decide to switch to the BOM (per other private discussion).

Author: Getting back to this discussion, I am hesitating about adding it. 
By using our bom, we will get rid of only one dependency, that is:
```
  <dependency>
      <groupId>com.datastax.oss.quarkus</groupId>
      <artifactId>cassandra-quarkus-client</artifactId>
      <version>${project.version}</version>
    </dependency>
```
by replacing it with:
```
<dependencyManagement>
    <dependencies>
      <dependency>
        <groupId>io.quarkus</groupId>
        <artifactId>cassandra-quarkus-bom</artifactId>
        <version>${project.version}</version>
        <type>pom</type>
        <scope>import</scope>
      </dependency>
    </dependencies>
  </dependencyManagement>
```
I am not sure if that is a good solution. I imagine that not all users are familiar with boms. Also, other quick start guides are using direct dependency to extension, instead of the bom (they are using bom internally in the extension as well)

Reviewer 2: Frankly I think these boms are a bit overkill. They contain 3/4 artifacts each. That's not enough to justify their presence.

Reviewer 2: Leaving as is for now.

The translation is done by quarkus-resteasy extension.
[source,java]
----
public class FruitDto {

  private String name;
  private String description;

  public FruitDto() {}

  public FruitDto(String name, String description) {
    this.name = name;
    this.description = description;
  }
  // getters and setters omitted for brevity
}
----

One important thing to note is that having a default constructor is required by the JSON serialization layer.

== Configuring the Cassandra database
The main two properties to configure are `contact-points`, to access the Cassandra database, and `local-datacenter`, which is required by the driver.

A sample configuration should look like this:

[source,properties]
----
# configure the Cassandra client for a replica set of two nodes
quarkus.cassandra.contact-points={cassandra_ip}:9042
quarkus.cassandra.load-balancing-policy.local-datacenter={dc_name}
----

In this example, we are using a single instance running on localhost:

[source,properties]
----
# configure the Cassandra client for a single instance on localhost
quarkus.cassandra.contact-points=127.0.0.1:9042
quarkus.cassandra.load-balancing-policy.local-datacenter=datacenter1
----

> You can configure all Java driver settings using `application.conf` or `application.json` files.

Reviewer: Maybe specify where these files should be located.

They will be passed automatically to the underlying driver configuration mechanism.
Settings defined in `application.properties` with the `quarkus.cassandra` prefix will have priority over settings defined in `application.conf` or `application.json`.

To see the full list of settings, please refer to the {reference-conf-url}[reference.conf].

== Running a Cassandra Database
By default, `CassandraClient` is configured to access a local Cassandra database on port 9042 (the default Cassandra port).
If you have a local running database on this port, you should assert that:

> `quarkus.cassandra.load-balancing-policy.local-datacenter` setting is equal to dc of your local cassandra instance.

Reviewer: Maybe we should specify that the value can be found with `SELECT data_center FROM system.local`, that might not be obvious for users not familiar with Cassandra.

* Apache Maven 3.5.3+
* Cassandra installed or Docker installed

== Architecture

The application built in this guide is quite simple: the user can add elements in a list using a form and the list is updated.

All the information between the browser and the server is formatted as JSON.

The elements are stored in the Cassandra Database.

== Solution

We recommend that you follow the instructions in the next sections and create the application step by step.
However, you can go right to the completed example.

The solution is located in the `quickstart` {quickstart-clone-url}[directory].

== Creating the Maven project

First, we need a new project. Create a new maven project and copy the `pom.xml` file that is present in the `quickstart`.

The `pom.xml` is importing the RESTEasy/JAX-RS, JSON-B, Context Propagation and Cassandra Client extensions.
We will be building a REST application that is using DataStax Object Mapper. It will simplify the Data Access Layer code.

The most important part of the `pom.xml` is adding the `cassandra-quarkus` extension:
[source,xml]
----
<dependency>
  <groupId>com.datastax.oss.quarkus</groupId>
  <artifactId>cassandra-quarkus-client</artifactId>

>   <version>${project.version}</version>

Reviewer 3: This isn't quite what we want, is it?  The user's project will shift versions as needed based on their development... I don't think we want to tie the version of the Quarkus extension to that.

</dependency>
----

== Creating JSON REST service
In this example, we will create an application to manage a list of fruits.

First, let's create the `Fruit` bean as follows:
[source,java]
----
@Entity
public class Fruit {

  @ClusteringColumn private String name;
  private String description;
  @PartitionKey private String id;

  public Fruit() {}

  public Fruit(String name, String description, String id) {
    this.name = name;
    this.description = description;
    this.id = id;
  }
  // getters, setters, hashCode and equals omitted for brevity
}
----

> We are using ObjectMapper, due to that fact the class is annotated with an `@Entity`.

Reviewer 3: Perhaps "ObjectMapper" here should be a pointer to https://docs.datastax.com/en/developer/java-driver/4.6/manual/mapper/ or other relevant docs to explain what we mean?

Author: I was referring to DataStax Object Mapper, link to docs added.

The last missing piece is the REST API that will expose GET and POST methods:
[source,java]
----
@Path("/fruits")
@Produces(MediaType.APPLICATION_JSON)
@Consumes(MediaType.APPLICATION_JSON)
public class FruitResource {

  private static final String STORE_NAME = "acme";
  @Inject FruitService fruitService;

  @GET
  public List<Fruit> list() {
    return fruitService.get(STORE_NAME);
  }

  @POST
  public void add(FruitDto fruit) {
    fruitService.save(covertFromDto(fruit));
  }

  private Fruit covertFromDto(FruitDto fruitDto) {
    return new Fruit(fruitDto.getName(), fruitDto.getDescription(), STORE_NAME);
  }
}
----

The list and add operations are executed for the `STORE_NAME` id. This is a partition key of the data model.
We can easily retrieve all rows from cassandra using the partition key. They will be sorted by the clustering key.
The `FruitResource` is using the `FruitService` that encapsulates the data access logic.

When creating the REST API we should not share the same entity object between REST API and data access layers.
They should not be coupled to allow evolving API independent of the storage layer.
This is the reason why the API is using the `FruitDto`.

> This class will be used by the quarkus to convert JSON to java object for client request and java object to JSON for the response.

Reviewer 3: Nit: s/used by the quarkus/used by Quarkus/

This translation is actually provided by REST Easy (https://resteasy.github.io/) so we might just want to say that here

== Configuring the Cassandra database
The main two properties to configure are `contact-points` to access to Cassandra and `local-datacenter` to use when connecting.

A sample configuration should look like this:

[source,properties]
----
# configure the Cassandra client for a replica set of two nodes
quarkus.cassandra.contact-points={cassandra_ip}:9042
quarkus.cassandra.load-balancing-policy.local-datacenter={dc_name}
----

In this example, we are using a single instance running on localhost:

[source,properties]
----
# configure the Cassandra client for a single instance on localhost
quarkus.cassandra.contact-points=127.0.0.1:9042
quarkus.cassandra.load-balancing-policy.local-datacenter=datacenter1
----

You can configure all java-driver settings using the `application.conf` or `application.json` files.
They will be passed automatically to the underlying java-driver config loader.
The settings from `application.properties` with `quarkus.cassandra` prefix have the priority.

To see the full list of settings, please refer to the {reference-conf-url}[reference.conf]

== Running a Cassandra Database

> As by default, `CassandraClient` is configured to access a local Cassandra database on port 9042 (the default Cassandra port).

Reviewer 3: s/As by default/By default/

If you have a local running database on this port, you should assert that:
`quarkus.cassandra.load-balancing-policy.local-datacenter` setting is equal to dc of your local cassandra instance.

If you want to use Docker to run a Cassandra database, you can use the following command to launch one:
[source,shell]
----
docker run \
   --name local-cassandra-instance \
   -p 7000:7000 \
   -p 7001:7001 \
   -p 7199:7199 \
   -p 9042:9042 \
   -p 9160:9160 \
   -p 9404:9404 \
   -d \
   launcher.gcr.io/google/cassandra3

> ----

Reviewer 3: Along the same lines as what we did in the quickstarter doc we should mention here that 9042 is required, all others are optional (but provide enhanced features like JMX monitoring of the C* instance etc.)

Next, you need to create the keyspace:
[source,shell]
----
docker exec -it local-cassandra-instance cqlsh -e "CREATE KEYSPACE IF NOT EXISTS k1 WITH replication = {'class':'SimpleStrategy', 'replication_factor':1}"
----

> And the Cassandra table that will be used by the quickstart application.

Reviewer 3: This should probably be "your sample application" rather than "the quickstart application" here

[source,shell]
----
docker exec -it local-cassandra-instance cqlsh -e "CREATE TABLE IF NOT EXISTS k1.fruit(id text, name text, description text, PRIMARY KEY((id), name))"

> ----

Reviewer 3: I wonder if it might read more cleanly to combine these into a single paragraph and show both code samples after the fact:

"Next you need to create the keyspace and table that will be used by your application:"
<sample1>
<sample2>

Author: It is split now already:
```

Next, you need to create the keyspace:
[source,shell]
----
docker exec -it local-cassandra-instance cqlsh -e "CREATE KEYSPACE IF NOT EXISTS k1 WITH replication = {'class':'SimpleStrategy', 'replication_factor':1}"
----

And the Cassandra table that will be used by the quickstart application.
[source,shell]
----
docker exec -it local-cassandra-instance cqlsh -e "CREATE TABLE IF NOT EXISTS k1.fruit(id text, name text, description text, PRIMARY KEY((id), name))"
```

Reviewer 2: He was asking the opposite: that you combine the snippets together. I think I agree it would read more cleanly.

> For the local Cassandra that is not running within docker, you need to execute cqlsh commands:

Reviewer 3: Perhaps something like "If you're running Cassandra locally you can execute the cqlsh commands directly:" might be cleaner here

This behavior must first be enabled by setting the `quarkus.cassandra.metrics.enabled` property to `true` in your `application.properties`.

The next step that you need to do is set explicitly which metrics should be enabled.
The `quarkus.cassandra.metrics.session-enabled` and `quarkus.cassandra.metrics.node-enabled` should be used for enabling metrics.
So for example to enable `session.connected-nodes` and `node.pool.open-connections` you should add the following settings to your
`application.properties`:
[source, properties]
----
quarkus.cassandra.metrics.enabled=true
quarkus.cassandra.metrics.session-enabled=connected-nodes
quarkus.cassandra.metrics.node-enabled=pool.open-connections
----

For the full list of available metrics, please refer to the {reference-conf-url}[reference.conf] and the `advanced.metrics` section.

So when you access the `/metrics` endpoint of your application you will have information about added metrics.

== Building a native executable

You can use the Cassandra client in a native executable.

You can build a native executable with the `mvn clean package -Pnative` command.

> Running it is as simple as executing `./target/quickstart-1.0.0-SNAPSHOT-runner`.

Reviewer 3: We've been describing this as creating a new app which happens to have the same content as the quickstart app rather than using the quickstart app itself.  Perhaps we should just tell the user to look for an app ending in "-runner" in the "target" directory here?

Author: When the user will follow this tutorial, he should copy the `pom.xml` from the quickstart. If he will do it, the name of the app will be quickstart - so the commands are proper

The elements are stored in the Cassandra Database.

== Solution

We recommend that you follow the instructions in the next sections and create the application step by step.
However, you can go right to the completed example.

The solution is located in the `quickstart` {quickstart-clone-url}[directory].

== Creating the Maven project

First, we need a new project. Create a new maven project and copy the `pom.xml` file that is present in the `quickstart`.

The `pom.xml` is importing the RESTEasy/JAX-RS, JSON-B, Context Propagation and Cassandra Client extensions.
We will be building a REST application that is using DataStax Object Mapper. It will simplify the Data Access Layer code.

The most important part of the `pom.xml` is adding the `cassandra-quarkus` extension:
[source,xml]
----
<dependency>
  <groupId>com.datastax.oss.quarkus</groupId>
  <artifactId>cassandra-quarkus-client</artifactId>
  <version>${quarkus.version}</version>
</dependency>
----

== Creating JSON REST service
In this example, we will create an application to manage a list of fruits.

First, let's create the `Fruit` bean as follows:
[source,java]
----
@Entity
public class Fruit {

>   @ClusteringColumn private String name;

Author: My idea was that `id` is a `store_name`:
`@PartitionKey private String id;`
and the REST API is inserting and retrieving data only for one particular store name. It is solving problem of fetch all. I can change the id column to `store_name`, wdyt?

>   @ClusteringColumn private String name;

Reviewer 2: But you need to change the quickstart project too, right? I agree that `store_name` is better, but since it's not related to this PR, how about doing it in a separate PR?

>   @ClusteringColumn private String name;

Author: Ok, I can do followup PR with that change.
I've created a JIRA task to not forget about it:
https://datastax-oss.atlassian.net/browse/JAVA-2767

>   @ClusteringColumn private String name;

Reviewer 2: Thank you!!

Also, the Fruit `id` is a `@PartitionKey` and `name` is a `@ClusteringColumn` so we are using the corresponding annotations
from the ObjectMapper. It will allow the Mapper to generate proper CQL queries underneath.

One important thing to note is that having a default constructor is required by the Object Mapper.

To leverage the Mapper logic in this app we need to create a DAO:
[source,java]
----
@Dao
public interface FruitDao {
  @Update
  void update(Fruit fruit);

  @Select
  PagingIterable<Fruit> findById(String id);
}
----

This class exposes operations that will be used in the REST service.

Finally, the Mapper itself:
[source,java]
----
@Mapper
public interface FruitMapper {
  @DaoFactory
  FruitDao fruitDao(@DaoKeyspace CqlIdentifier keyspace);
}
----
It is constructing the `FruitDao`.

Now create a `com.datastax.oss.quarkus.FruitService` that will be the business layer of our application and store/load the fruits from the Cassandra database.
[source,java]
----
@ApplicationScoped
public class FruitService {
  private final FruitDao dao;

  @Inject
  public FruitService(CqlSession session) {

>     dao = new FruitMapperBuilder(session).build().fruitDao(CqlIdentifier.fromCql("k1"));

Author: 1. I can create an application bean that is `FruitDao`, but I am not sure if it is needed. It will add some complexity to this example by adding one additional layer, wdyt?

2. Regarding using the session name from the config file, I am not sure if we should provide it.
- We can add one additional setting in the application.properties of our starter called keyspace, but it will restrict usage of the `CqlSession` to one keyspace. I think it may be too limiting.
- Right now, the problem of providing keyspace name is on the user - he needs to use any mechanism that he wants, to propagate it. The user when creating own quarkus app can use their own `@ConfigRoot` and propagate it, or have it hardcoded. 

I am not sure which direction is better, for spring we did configure keyspace at the start and restricted the `CqlSession` to one keyspace. Maybe we should wait for feedback before changing it?

>     dao = new FruitMapperBuilder(session).build().fruitDao(CqlIdentifier.fromCql("k1"));

Reviewer: Keyspace: right now it's hardcoded, so it's even less flexible. I think moving it to the configuration is good enough for this example. The only case it doesn't cover is having two `FruitService` beans that hit different keyspaces: that's possible, but doesn't seem like a very common use case.

Injecting `FruitDao`: I'm not sure if I would go that far, but we should at least inject the `FruitMapper`. A typical application can probably have more services (`StoreService`, `UserService`, etc), and in such a case you don't want to re-create a mapper instance for every service.

>     dao = new FruitMapperBuilder(session).build().fruitDao(CqlIdentifier.fromCql("k1"));

Author: 1. Keyspace from settings implemented in https://github.com/datastax/cassandra-quarkus/pull/9/commits/97821a137d3fee2823388eef83979271e242e30a and docs 
https://github.com/datastax/cassandra-quarkus/pull/9/commits/76a99f9f0f2ecb945b50a28ac70af674f3e606a3

2. Injecting `FrutiMapper`: implemented in https://github.com/datastax/cassandra-quarkus/pull/9/commits/ab57a565a7d5fd85e82d81a41a39c19d20b4720b and docs
https://github.com/datastax/cassandra-quarkus/pull/9/commits/cb0be06dc99f4762bcbaee47da3e9337deb9ae0e

  }

  public void save(Fruit fruit) {
    dao.update(fruit);
  }

  public List<Fruit> get(String id) {
    return dao.findById(id).all();
  }
}
----
Please note the construction of dao. It is using `CqlSession` injected automatically by the cassandra-quarkus extension.

The last missing piece is the REST API that will expose GET and POST methods:
[source,java]
----
@Path("/fruits")
@Produces(MediaType.APPLICATION_JSON)
@Consumes(MediaType.APPLICATION_JSON)
public class FruitResource {

  private static final String STORE_NAME = "acme";
  @Inject FruitService fruitService;

  @GET
  public List<Fruit> list() {

>     return fruitService.get(STORE_NAME);

Reviewer 2: I definitely don't understand this data model. Isn't `id` supposed to be the fruit id?

Author: It is listing all fruits withing STORE_ID partition key.

If you have a local running database on this port, you should assert that:
`quarkus.cassandra.load-balancing-policy.local-datacenter` setting is equal to dc of your local cassandra instance.

If you want to use Docker to run a Cassandra database, you can use the following command to launch one:
[source,shell]
----
docker run \
   --name local-cassandra-instance \
   -p 7000:7000 \
   -p 7001:7001 \
   -p 7199:7199 \
   -p 9042:9042 \
   -p 9160:9160 \
   -p 9404:9404 \
   -d \
   launcher.gcr.io/google/cassandra3
----
Note that only the 9042 port is required. All others all optional but provide enhanced features like JMX monitoring of the C* instance.

Next, you need to create the keyspace:
[source,shell]
----
docker exec -it local-cassandra-instance cqlsh -e "CREATE KEYSPACE IF NOT EXISTS k1 WITH replication = {'class':'SimpleStrategy', 'replication_factor':1}"
----

And the Cassandra table that will be used by your quickstart application.
[source,shell]
----
docker exec -it local-cassandra-instance cqlsh -e "CREATE TABLE IF NOT EXISTS k1.fruit(id text, name text, description text, PRIMARY KEY((id), name))"
----

> For the local Cassandra that is not running within docker, you need to execute cqlsh commands:

Reviewer 2: ```suggestion
If you're running Cassandra locally you can execute the cqlsh commands directly:
```

In the `src/main/resources/META-INF/resources` directory, add a `fruits.html` file with the content from this {fruits-html-url}[fruits.html] file in it.

You can now interact with your REST service:

* start Quarkus with `mvn clean quarkus:dev`
* open a browser to `http://localhost:8080/fruits.html`
* add new fruits to the list via the form

== Connection Health Check

If you are using the `quarkus-smallrye-health` extension, `cassandra-quarkus` will automatically add a readiness health check
to validate the connection to the cluster.

So when you access the `/health/ready` endpoint of your application you will have information about the connection validation status.

This behavior can be disabled by setting the `quarkus.cassandra.health.enabled` property to `false` in your `application.properties`.

== Metrics

If you are using the `quarkus-smallrye-metrics` extension, `cassandra-quarkus` can provide metrics about CqlSession and Cassandra nodes.
This behavior must first be enabled by setting the `quarkus.cassandra.metrics.enabled` property to `true` in your `application.properties`.

The next step that you need to do is set explicitly which metrics should be enabled.

> The `quarkus.cassandra.metrics.session-enabled` and `quarkus.cassandra.metrics.node-enabled` should be used for enabling metrics.

Reviewer 2: ```suggestion
The `quarkus.cassandra.metrics.session-enabled` and `quarkus.cassandra.metrics.node-enabled` properties should be used for enabling metrics; the former should contain a list of session-level metrics to enable, while the latter should contain a list of node-level metrics to enable. Both properties accept a comma-separated list of valid metric names.
```

So for example to enable `session.connected-nodes` and `node.pool.open-connections` you should add the following settings to your
`application.properties`:
[source, properties]
----
quarkus.cassandra.metrics.enabled=true

> quarkus.cassandra.metrics.session-enabled=connected-nodes

Reviewer 2: What's the exact syntax to use if there is more than one metric enabled? Maybe you should add an example of this.

Author: done

quarkus.cassandra.metrics.node-enabled=pool.open-connections
----

For the full list of available metrics, please refer to the {reference-conf-url}[reference.conf] and the `advanced.metrics` section.

> So when you access the `/metrics` endpoint of your application you will have information about added metrics.

Reviewer 2: ```suggestion
When metrics are properly enabled and when you access the `/metrics` endpoint of your application, you will see metric reports for all enabled metrics.
```

== Building a native executable

You can use the Cassandra client in a native executable.

You can build a native executable with the `mvn clean package -Pnative` command.

Running it is as simple as executing `./target/quickstart-1.0.0-SNAPSHOT-runner`.

You can then point your browser to `http://localhost:8080/fruits.html` and use your application.

== Conclusion

> Accessing a Cassandra database from a Cassandra Client is easy with Quarkus as it provides configuration and native support for it.

Reviewer 2: ```suggestion
Accessing a Cassandra database from a client application is easy with Quarkus and the Cassandra extension, which provides configuration and native support for the DataStax Java driver for Apache Cassandra.
```

= Quarkus - Using the Cassandra Client

Apache Cassandra® is a free and open-source, distributed, wide column store, NoSQL database management system designed
to handle large amounts of data across many commodity servers, providing high availability with no single point of failure.

In this guide, we will see how you can get your REST services to use a Cassandra database.

== Prerequisites

To complete this guide, you need:

* an IDE
* JDK 1.8+ installed with `JAVA_HOME` configured appropriately
* GraalVM installed with `GRAALVM_HOME` configured appropriately if you want to use the native mode.
* Apache Maven 3.5.3+
* Cassandra installed or Docker installed

== Architecture

The application built in this guide is quite simple: the user can add elements in a list using a form and the list is updated.

All the information between the browser and the server is formatted as JSON.

The elements are stored in the Cassandra Database.

== Solution

We recommend that you follow the instructions in the next sections and create the application step by step.
However, you can go right to the completed example.

> The solution is located in the `quickstart` {quickstart-clone-url}[directory].

Author: <img width="737" alt="Screenshot 2020-05-14 at 10 28 11" src="https://user-images.githubusercontent.com/2332235/81911448-9d699c80-95cd-11ea-8221-cac2d68c361f.png">
It is generating for me

> The solution is located in the `quickstart` {quickstart-clone-url}[directory].

Reviewer 2: It is rendered correctly for me too.

> The solution is located in the `quickstart` {quickstart-clone-url}[directory].

Reviewer 3: It is now for me as well too... between this and the "only PDF" comment above I'm starting to wonder what was up with my runs of this thing under the prior commits.  I ran it multiple times (with cleans in between) and saw consistent results... but I agree this is fixed now.

[source,java]
----
@Path("/fruits")
@Produces(MediaType.APPLICATION_JSON)
@Consumes(MediaType.APPLICATION_JSON)
public class FruitResource {

  private static final String STORE_NAME = "acme";
  @Inject FruitService fruitService;

  @GET
  public List<Fruit> list() {
    return fruitService.get(STORE_NAME);
  }

  @POST
  public void add(FruitDto fruit) {
    fruitService.save(covertFromDto(fruit));
  }

  private Fruit covertFromDto(FruitDto fruitDto) {
    return new Fruit(fruitDto.getName(), fruitDto.getDescription(), STORE_NAME);
  }
}
----

The `list` and `add` operations are executed for the `STORE_NAME` id. This is a partition key of the data model.
We can easily retrieve all rows from cassandra using the partition key. They will be sorted by the clustering key.
`FruitResource` is using `FruitService` which encapsulates the data access logic.

When creating the REST API we should not share the same entity object between REST API and data access layers.
They should not be coupled to allow the API to evolve independently of the storage layer.
This is the reason why the API is using a `FruitDto` class.

> This class will be used by Quarkus to convert JSON to java object for client request and java object to JSON for the response.

Reviewer 3: I agree that this is the right decision architecturally... but it's not what we're doing right now.

More specifically we are doing the right thing in the POST handler; we're expecting a JSON rep of a FruitDto and working with it as such.  But the handler for the GET makes no use of FruitDto at all.  And there's nothing in Quarkus or REST Easy that knows that we should render Fruit as a FruitDto instead (or knows how to translate one into the other).

To validate, note what we get with a simple curl request to the Web service:

```
[mersault@localhost ~]$ curl http://localhost:8080/fruits
[{"description":"foo","id":"acme","name":"bar"},{"description":"bar","id":"acme","name":"foo"}]
```

The "id" field is not in FruitDto... but it is in Fruit.  What's actually happening here is that the collection of Fruit objects returned from FruitResource.list() is being serialized to JSON directly.  But, as the comment suggests, we should be using FruitDto for this operation.

We need something like the following instead:

```java
  @GET
  public List<FruitDto> list() {
    return fruitService.get(STORE_NAME).stream()
            .map(fruit -> new FruitDto(fruit.getName(), fruit.getDescription()))
            .collect(Collectors.toList());
  }
```

Note the output of the curl request with this change:

```
[mersault@localhost ~]$ curl http://localhost:8080/fruits
[{"description":"foo","name":"bar"},{"description":"bar","name":"foo"}]
```

quarkus.cassandra.contact-points={cassandra_ip}:9042
quarkus.cassandra.load-balancing-policy.local-datacenter={dc_name}
----

In this example, we are using a single instance running on localhost:

[source,properties]
----
# configure the Cassandra client for a single instance on localhost
quarkus.cassandra.contact-points=127.0.0.1:9042
quarkus.cassandra.load-balancing-policy.local-datacenter=datacenter1
----

You can configure all Java driver settings using `application.conf` or `application.json` files.
They need to be located in the resources directory of your application.
They will be passed automatically to the underlying driver configuration mechanism.
Settings defined in `application.properties` with the `quarkus.cassandra` prefix will have priority over settings defined in `application.conf` or `application.json`.

To see the full list of settings, please refer to the {reference-conf-url}[reference.conf].

== Running a Cassandra Database
By default, `CassandraClient` is configured to access a local Cassandra database on port 9042 (the default Cassandra port).
If you have a local running database on this port, you should assert that:

> `quarkus.cassandra.load-balancing-policy.local-datacenter` setting is equal to dc of your local cassandra instance.

Reviewer 3: Nit: "setting is equal to the DC of your local Cassandra instance".  We capitalized "Cassandra" everywhere else in this paragraph so we probably should here.

[source,shell]
----
docker run \
   --name local-cassandra-instance \
   -p 7000:7000 \
   -p 7001:7001 \
   -p 7199:7199 \
   -p 9042:9042 \
   -p 9160:9160 \
   -p 9404:9404 \
   -d \
   launcher.gcr.io/google/cassandra3
----
Note that only the 9042 port is required. All others all optional but provide enhanced features like JMX monitoring of the C* instance.

Next you need to create the keyspace and table that will be used by your application:
[source,shell]
----
docker exec -it local-cassandra-instance cqlsh -e "CREATE KEYSPACE IF NOT EXISTS k1 WITH replication = {'class':'SimpleStrategy', 'replication_factor':1}"
docker exec -it local-cassandra-instance cqlsh -e "CREATE TABLE IF NOT EXISTS k1.fruit(id text, name text, description text, PRIMARY KEY((id), name))"
----

If you're running Cassandra locally you can execute the cqlsh commands directly:
[source,shell]
----
cqlsh -e "CREATE KEYSPACE IF NOT EXISTS k1 WITH replication = {'class':'SimpleStrategy', 'replication_factor':1}
cqlsh -e "CREATE TABLE IF NOT EXISTS k1.fruit(id text, name text, description text, PRIMARY KEY((id), name))

> ----

Reviewer 3: Here we _can_ point people to ```cqlsh -f setup_db.cql```, right?  Might be simpler than having to show the statements directly inline again.

Author: I disagree. It is simpler to have those statements in the docs. Otherwise, the user will need to lookup another file.

Reviewer 3: I"m not sure I agree, but I don't feel terribly strongly about it... so let's keep them as they are.

quarkus.cassandra.contact-points={cassandra_ip}:9042
quarkus.cassandra.load-balancing-policy.local-datacenter={dc_name}
----

In this example, we are using a single instance running on localhost:

[source,properties]
----
# configure the Cassandra client for a single instance on localhost
quarkus.cassandra.contact-points=127.0.0.1:9042
quarkus.cassandra.load-balancing-policy.local-datacenter=datacenter1
----

You can configure all Java driver settings using `application.conf` or `application.json` files.
They need to be located in the resources directory of your application.
They will be passed automatically to the underlying driver configuration mechanism.
Settings defined in `application.properties` with the `quarkus.cassandra` prefix will have priority over settings defined in `application.conf` or `application.json`.

To see the full list of settings, please refer to the {reference-conf-url}[reference.conf].

== Running a Cassandra Database
By default, `CassandraClient` is configured to access a local Cassandra database on port 9042 (the default Cassandra port).
If you have a local running database on this port, you should assert that:

> `quarkus.cassandra.load-balancing-policy.local-datacenter` setting is equal to dc of your local Cassandra instance.

Reviewer 3: Nit: since the name "datacenter" is used in the config probably prefer "equal to the datacenter of your local Cassandra instance" here.

This value can be found with `SELECT data_center FROM system.local` query.

If you want to use Docker to run a Cassandra database, you can use the following command to launch one:
[source,shell]
----
docker run \
   --name local-cassandra-instance \
   -p 7000:7000 \
   -p 7001:7001 \
   -p 7199:7199 \
   -p 9042:9042 \
   -p 9160:9160 \
   -p 9404:9404 \
   -d \
   launcher.gcr.io/google/cassandra3
----

> Note that only the 9042 port is required. All others all optional but provide enhanced features like JMX monitoring of the C* instance.

Reviewer 3: Semi-nit: we've used the actual name "Cassandra" everywhere else, probably want that here instead of "C*"
